core/chains/chains.go

```diff
@@ -0,0 +1,137 @@
+
+// Chain represents a blockchain with a unique Chain ID
+type Chain struct {
+	id      *big.Int
+	config  ChainSpecificConfig
+	logOnce sync.Once
+}
+
+func (c *Chain) setChainID(id int64) {
+	c.id = big.NewInt(id)
+}
+
+func (c *Chain) ID() *big.Int {
+	return c.id
+}
+
+func (c *Chain) Config() ChainSpecificConfig {
+	if !c.config.set {
+		c.logOnce.Do(func() {
+			logger.Warnf("chain with ID %s does not have a chain-specific config, using fallback config instead", c.ID())
+		})
+		return FallbackConfig
+	}
+	return c.config
+}
+
+// IsArbitrum returns true if the chain is arbitrum mainnet or testnet
+func (c *Chain) IsArbitrum() bool {
+	if c == nil {
+		return false
+	}
+	return c.ID().Cmp(ArbitrumMainnet.ID()) == 0 || c.ID().Cmp(ArbitrumRinkeby.ID()) == 0
+}
+
+// IsOptimism returns true if the chain is optimism mainnet or testnet
+func (c *Chain) IsOptimism() bool {
+	if c == nil {
+		return false
+	}
+	return c.ID().Cmp(OptimismMainnet.ID()) == 0 || c.ID().Cmp(OptimismKovan.ID()) == 0
+}
+
+// IsL2 returns true if this chain is an L2 chain, notably that the block
+// numbers used for log searching are different from calling block.number
+func (c *Chain) IsL2() bool {
+	return c.IsOptimism() || c.IsArbitrum()
+}
+
+var chains map[int64]*Chain
+var (
+	EthMainnet      = new(Chain)
+	EthRinkeby      = new(Chain)
+	EthGoerli       = new(Chain)
+	EthKovan        = new(Chain)
+	OptimismMainnet = new(Chain)
+	OptimismKovan   = new(Chain)
+	ArbitrumMainnet = new(Chain)
+	ArbitrumRinkeby = new(Chain)
+	BSCMainnet      = new(Chain)
+	HecoMainnet     = new(Chain)
+	FantomMainnet   = new(Chain)
+	FantomTestnet   = new(Chain)
+	PolygonMainnet  = new(Chain)
+	PolygonMumbai   = new(Chain)
+	XDaiMainnet     = new(Chain)
+	RSKMainnet      = new(Chain)
+	AvalancheFuji   = new(Chain)
+)
+
+func init() {
+	chains = make(map[int64]*Chain)
+
+	chains[1] = EthMainnet
+	chains[4] = EthRinkeby
+	chains[5] = EthGoerli
+	chains[42] = EthKovan
+	chains[10] = OptimismMainnet
+	chains[69] = OptimismKovan
+	chains[42161] = ArbitrumMainnet
+	chains[421611] = ArbitrumRinkeby
+	chains[56] = BSCMainnet
+	chains[128] = HecoMainnet
+	chains[250] = FantomMainnet
+	chains[4002] = FantomTestnet
+	chains[137] = PolygonMainnet
+	chains[80001] = PolygonMumbai
+	chains[100] = XDaiMainnet
+	chains[30] = RSKMainnet
+	chains[43113] = AvalancheFuji
+
+	for id, chain := range chains {
+		chain.setChainID(id)
+	}
+
+	setConfigs()
+}
+
+var chainsMu sync.RWMutex
+
+// ChainFromID returns the chain for the given ID
+// If no chain is found, creates a new one and returns that
+func ChainFromID(id *big.Int) *Chain {
+	if !id.IsInt64() {
+		panic("chain IDs larger than the max 64 bit integer are not currently supported")
+	}
+	// Use rlock for the hot path
+	chainsMu.RLock()
+	chain, exists := chains[id.Int64()]
+	chainsMu.RUnlock()
+	if exists {
+		return chain
+	}
+	chainsMu.Lock()
+	defer chainsMu.Unlock()
+	chain, exists = chains[id.Int64()]
+	if exists {
+		return chain
+	}
```

Comment:
i guess this fragment is duplicated because of different lock usage?
```
	chain, exists = chains[id.Int64()]
	if exists {
		return chain
	}
```
still, seems like it could be just done once easily, from what i can see

Reply:
Seems like this is confusing to both you and JB, I will use the simpler version instead.

Reply:
The original intention was to make the common case (read) use an RLock.

In the uncommon case (set when unset) we use a write Lock, but we need to read again inside the write lock in case there are multiple writers.

This optimisation is probably overkill and confusing, so I will remove it.

Reply:
`we need to read again inside the write lock in case there are multiple writers.` ah, right - so that we make sure the writes will not override each other in the extreme case. 
Yeah maybe a comment or extracting this to a separate method would help in readability